view/src/storage.rs

```diff
@@ -666,6 +697,97 @@ impl Storage {
+        let mut tx = self.pool.begin().await?;
+
+        let note_commitment = note.commit().0.to_bytes().to_vec();
+        let address = note.address().to_vec();
+        let amount = u64::from(note.amount()) as i64;
+        let asset_id = note.asset_id().to_bytes().to_vec();
+        let blinding_factor = note.note_blinding().to_bytes().to_vec();
+
+        sqlx::query!(
+            "INSERT INTO notes
+                    (
+                        note_commitment,
+                        address,
+                        amount,
+                        asset_id,
+                        blinding_factor
+                    )
+                    VALUES
+                    (?, ?, ?, ?, ?)",
+            note_commitment,
+            address,
+            amount,
+            asset_id,
+            blinding_factor,
+        )
+        .execute(&mut tx)
+        .await?;
+
+        tx.commit().await?;
+
+        Ok(())
+    }
+
+    /// Return advice about note contents for use in scanning.
+    ///
+    /// Given a list of note commitments, this method checks whether any of them
+    /// correspond to notes that have been recorded in the database but not yet
+    /// observed during scanning.
+    pub async fn scan_advice(
+        &self,
+        note_commitments: Vec<note::Commitment>,
+    ) -> anyhow::Result<BTreeMap<note::Commitment, Note>> {
+        if note_commitments.is_empty() {
+            return Ok(BTreeMap::new());
+        }
+
+        let rows = sqlx::query(
+            format!(
+                "SELECT notes.address,
+                        notes.amount,
+                        notes.asset_id,
+                        notes.blinding_factor
+                FROM notes
+                LEFT OUTER JOIN spendable_notes ON notes.note_commitment = spendable_notes.note_commitment
+                WHERE (spendable_notes.note_commitment IS NULL) AND (notes.note_commitment IN ({}))",
+                note_commitments
+                    .iter()
+                    .map(|cm| format!("x'{}'", hex::encode(cm.0.to_bytes())))
+                    .collect::<Vec<_>>()
+                    .join(", ")
+            )
+            .as_str(),
+        )
+        .fetch_all(&self.pool)
+        .await?;
+
+        let mut notes = BTreeMap::new();
+        for row in rows {
+            let address = Address::try_from(row.get::<&[u8], _>("address"))?;
+            let amount = (row.get::<i64, _>("amount") as u64).into();
+            let asset_id = asset::Id(Fq::from_bytes(
+                row.get::<&[u8], _>("asset_id")
+                    .try_into()
+                    .expect("32 bytes"),
+            )?);
+            let blinding_factor = Fq::from_bytes(
+                row.get::<&[u8], _>("blinding_factor")
+                    .try_into()
+                    .expect("32 bytes"),
+            )?;
+
+            let note =
+                Note::from_parts(address, Value { amount, asset_id }, blinding_factor).unwrap();
```

Comment:
Followup: refactor into a wrapper type with `sqlx::FromRow`

This is copied out of the `FromRow` impl for `SpendableNoteRecord`; we can't implement `FromRow` on `Note` directly, because of orphan rules, but we could probably clean this up with a newtype wrapper.